File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D24.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y15D24(input: String) : Day {

    private val lines = input.lines()

    private fun List<Int>.product() = this.fold(1L, Long::times)

    private fun uniqueCombos(validCombos: List<List<Int>>, taken: List<List<Int>>, uniqueGroups: Int): List<Int>? {
        if (validCombos.isEmpty()) {
            return null
        } else {
            for (i in validCombos.indices) {
                if (validCombos[i].intersect(taken.flatten().toSet()).isEmpty()) {
                    return if (uniqueGroups > 3) {
                        val subUnique = uniqueCombos(
                            validCombos.drop(i),
                            taken + listOf(validCombos[i]),
                            uniqueGroups - 1
                        )
                        if (subUnique == null) continue else taken.first()
                    } else {
                        taken.first()
                    }
                }
            }
            return null
        }
    }

    private val weights = lines
        .map(String::toInt)
        .sortedDescending()

    data class Package(val weight: Int, val range: IntRange)

    fun solve(numberOfGroups: Int): Long {
        val groupWeight = weights.sum() / numberOfGroups

        val packages = weights.mapIndexed { i, weight ->
            val lowEnd = groupWeight - weights.drop(i).sum()
            val highEnd = groupWeight - weight
            Package(weight, lowEnd..highEnd)
        }

        var combos = packages
            .filter { 0 in it.range }
            .map { listOf(it.weight) }

        var validCombos = mutableListOf<List<Int>>().apply {
            addAll(combos.filter { it.sum() == groupWeight })
        }

        // do a pass
        while (combos.isNotEmpty()) {
            val newCombos = mutableListOf<List<Int>>()

            val startIndex = packages.indexOf(packages.find { it.weight == combos.first().last() }) + 1

            var i = startIndex
            loop@while (i < packages.size) {
                for (combo in combos) {
                    if (combo.last() <= packages[i].weight) {
                        i++
                        continue@loop
                    }
                    if (combo.sum() in packages[i].range) {
                        val latestCombo = combo + packages[i].weight
                        if (latestCombo.sum() == groupWeight) {
                            validCombos.add(latestCombo)
                        } else {
                            newCombos.add(latestCombo)
                        }
                    }
                }
                i++
            }
            combos = newCombos
            if (validCombos.size >= numberOfGroups - 1) {
                validCombos = validCombos.sortedBy { it.product() }.toMutableList()
                val unique = uniqueCombos(validCombos.drop(1), listOf(validCombos.first()), 3)
                if (unique != null) {
                    return unique.product()
                }
            }
        }
        return -1L
    }


    override fun part1() = solve(3)

    override fun part2() = solve(4)
}

fun main() = Day.runDay(Y15D24::class)

//    Class creation: 19ms
//    Part 1: 11846773891 (230ms)
//    Part 2: 80393059 (2ms)
//    Total time: 253ms

File: rust/advent/src/bin/2015/y15d24.rs
use std::ops::RangeInclusive;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use rustc_hash::FxHashSet;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = Vec<i64>;
type Output = i64;

#[derive(Debug, Clone)]
struct Package {
    weight: i64,
    range: RangeInclusive<i64>
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 24).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    input.lines()
        .map(|line| line.parse().unwrap())
        .sorted_unstable()
        .rev()
        .collect()
}

fn solve(weights: &Input, groups: i64) -> Output {
    let group_weight: i64 = weights.iter().sum::<i64>() / groups;
    let packages: Vec<Package> = weights.iter().enumerate()
        .map(|(i, &weight)| {
            let low_end = group_weight - &weights[i..].iter().sum();
            let high_end = group_weight - weight;
            Package { weight, range: low_end..=high_end }
        })
        .collect();

    let mut combos: Vec<Vec<i64>> = packages.iter()
        .filter(|package| package.range.contains(&0))
        .map(|package| vec![package.weight])
        .collect();
    let mut valid_combos = combos.iter()
        .filter(|&combo| combo.iter().sum::<i64>() == group_weight)
        .cloned()
        .collect_vec();

    while !combos.is_empty() {
        let mut new_combos = Vec::new();
        let last_first = *combos[0].last().unwrap();
        let mut i = 1 + packages.iter()
            .position(|package| package.weight == last_first)
            .unwrap();
        'middle:while i < packages.len() {
            for combo in combos.iter() {
                let package = packages.get(i).unwrap();
                if *combo.last().unwrap() <= package.weight {
                    i += 1;
                    continue 'middle;
                }
                if package.range.contains(&combo.iter().sum()) {
                    let latest_combo = combo.iter()
                        .copied()
                        .chain(std::iter::once(package.weight))
                        .collect_vec();
                    if latest_combo.iter().sum::<i64>() == group_weight {
                        valid_combos.push(latest_combo);
                    } else {
                        new_combos.push(latest_combo);
                    }
                }
            }
            i += 1;
        }
        std::mem::swap(&mut combos, &mut new_combos);
        if valid_combos.len() >= groups as usize - 1 {
            valid_combos.sort_unstable_by_key(|it| {
                it.iter().fold(1, |acc, &w| acc * w)
            });
            let taken: FxHashSet<i64> = valid_combos[0].iter().copied().collect();
            for combo in valid_combos.iter() {
                if combo.iter().all(|weight| !taken.contains(weight)) {
                    return taken.into_iter().fold(1, |acc, w| acc * w)
                }
            }
        }
    }
    unreachable!()
}

fn part1(input: &Input) -> Output {
    solve(input, 3)
}

fn part2(input: &Input) -> Output {
    solve(input, 4)
}

#[test]
fn default() {
    let input = get_input(15, 24).unwrap();
    let input = parse_input(&input);
    assert_eq!(11846773891, part1(&input));
    assert_eq!(80393059, part2(&input));
}

// Input parsed (17μs)
// 1. 11846773891 (23ms)
// 2. 80393059 (1ms)
// Total: 25ms